File: NodeJS/google_serp_scraper_hasdata.js
const axios = require('axios');
const fs = require('fs');

const apiKey = 'YOUR-API-KEY';

const query = 'site:hasdata.com inurl:blog';
const location = 'Austin,Texas,United States';
const deviceType = 'desktop';
const numResults = 100;

const baseUrl = 'https://api.hasdata.com/scrape/google/serp';

const params = {
  q: query,
  location: location,
  deviceType: deviceType,
  num: numResults,
};

const headers = {
  'Content-Type': 'application/json',
  'x-api-key': apiKey,
};

async function fetchSerp() {
  try {
    const response = await axios.get(baseUrl, { headers, params });
    const data = response.data.organicResults || [];
    const urls = data
      .filter(entry => entry.link)
      .map(entry => entry.link);

    const outputFile = 'serp.json';
    fs.writeFileSync(outputFile, JSON.stringify(urls, null, 2), 'utf-8');

    console.log(`Saved ${urls.length} URLs to ${outputFile}`);
  } catch (error) {
    console.error('Error fetching SERP data:', error.message);
  }
}

fetchSerp();

File: Python/google_serp_scraper_hasdata.py
import requests
import json

api_key = "YOUR-API-KEY"

query = "site:hasdata.com inurl:blog"
location = "Austin,Texas,United States"
device_type = "desktop"
num_results = 100

base_url = "https://api.hasdata.com/scrape/google/serp"

params = {
    "q": query,
    "location": location,
    "deviceType": device_type,
    "num": num_results
}

headers = {
    "Content-Type": "application/json",
    "x-api-key": api_key
}

response = requests.get(base_url, headers=headers, params=params)
response.raise_for_status()

data = response.json().get("organicResults", [])
urls = [entry["link"] for entry in data if "link" in entry]

output_file = "serp.json"
with open(output_file, "w", encoding="utf-8") as f:
    json.dump(urls, f, indent=2)

print(f"Saved {len(urls)} URLs to {output_file}")
